presto-spi/src/main/java/com/facebook/presto/spi/block/BlockUtil.java

```diff
@@ -35,11 +33,27 @@ private BlockUtil()
     {
     }
 
-    static void checkValidPositions(List<Integer> positions, int positionCount)
+    static void checkValidPositions(int[] positions, int offset, int length, int positionCount)
     {
-        Set<Integer> invalidPositions = positions.stream().filter(position -> position >= positionCount).collect(toSet());
-        if (!invalidPositions.isEmpty()) {
-            throw new IllegalArgumentException("Invalid positions " + invalidPositions + " in block with " + positionCount + " positions");
+        checkValidPositionsArray(positions, offset, length);
+
+        for (int i = offset; i < offset + length; ++i) {
+            int position = positions[i];
+            if (position > positionCount) {
+                throw new IllegalArgumentException("Invalid position " + position + " in block with " + positionCount + " positions");
+            }
+        }
+    }
+
+    static void checkValidPositionsArray(int[] positions, int offset, int length)
+    {
+        requireNonNull(positions, "positions array is null");
+        if (offset < 0 || offset > positions.length) {
```

Comment:
What if `offset == positions.length` and `length == 0`?

Reply:
There wasn't any requirement before that length has to be non-zero to so I haven't introduced one either.

Reply:
What I meant was if `offset == positions.length` and `length == 0`, it will pass the check. But it doesn't really make sense to have `offset == positions.length`. So probably might be good to use `>=` in the check?

Reply:
By `offset == positions.length` and `length == 0` I mean: `copyPositions(new int[0], 1, 0)`

Reply:
OK sorry I didn't notice `new int[0]`. Sure, if that's a valid case, we can leave that.

Reply:
> Therefore it is legal to call `copyPositions(new int[0], 0, 0)`. It was legal before to call it with empty list too.

I think it's logical for this to be legal, but _generally_ we should avoid creating empty pages.

Reply:
I don't mean, however, that avoiding empty pages should be a concern of `copyPositions` (in fact, since caller would need to special-handle no-page-for-empty-page, it helps nothing to handle this here too). Thus, no change is requested.